### ISSUE TITLE
Service scope seems to be affecting precedence for tagged services, causing container bootstrap to fail

### ISSUE THREAD

**Issue by AshleyDawson** [symfony/symfony#28720]
**Symfony version(s) affected**: 4.1.6

**Description**  
Dissimilar service scope when tagging services seems to break the precedence of service injection in the container cache bootstrap file. 

**How to reproduce**  
Tag a `public: true` service into a `public: false` service. E.g.

```
    app.form.post_processor.post_processor_registry:
        public: false
        class: App\Bundle\FormBundle\PostProcessor\PostProcessorRegistry

    App\Bundle\FormBundle\PostProcessor\NotifyPostProcessor:
        public: true
        arguments:
            - "@app.mailer.message_builder"
            - "@app.mailer.message_sender"
        tags:
            - { name: app.form.post_processor }
```

With the compiler pass:

```
<?php

namespace App\Bundle\FormBundle\DependencyInjection\Compiler;

use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
use Symfony\Component\DependencyInjection\ContainerBuilder;
use Symfony\Component\DependencyInjection\Reference;

class PostProcessorCompilerPass implements CompilerPassInterface
{
    /**
     * {@inheritdoc}
     */
    public function process(ContainerBuilder $container)
    {
        if (! $container->has('app.form.post_processor.post_processor_registry')) {
            return;
        }

        $definition = $container
            ->findDefinition('app.form.post_processor.post_processor_registry')
        ;

        $taggedProviders = $container
            ->findTaggedServiceIds('app.form.post_processor')
        ;

        foreach ($taggedProviders as $id => $tags) {
            $definition->addMethodCall('add', [
                new Reference($id)
            ]);
        }
    }
}
```

This then produces the following bootstrap in `var/cache/dev/ContainerYZDvbSj/srcDevDebugProjectContainer.php`:

```
        $u->add(new \App\Bundle\FormBundle\PostProcessor\NotifyPostProcessor($g, $h));

        $u = new \App\Bundle\FormBundle\PostProcessor\PostProcessorRegistry();
```

Notice that `PostProcessorRegistry` is instantiated after `$u->add(...)`. This therefore causes the application to fail due to this issue in the service container bootstrap file. 

**Possible Solution**  
Workaround is to set all services to the same scope (e.g. all to `public`)

**Comment by AshleyDawson** [comment]
This is probably related to issue #28710

**Comment by nicolas-grekas** [comment]
Same as #28710: sharing the dumped xml container would greatly help (even privately, I'm available on Symfony-dev's Slack.)

**Comment by AshleyDawson** [comment]
Hi Nicolas, trying to sign into the Symfony dev slack channel, but the invite link is no longer active: https://symfony-devs.slack.com/join/shared_invite/enQtNDI4MDc0NjgwNDM0LWQ1ZDA3MzJhNTQxZjIxMzY5YzNlYTBjNjhiOGNhYTU4OTliMmQ4YmIxN2U3NDhkNmQ5ODU2OTc5ODQ2Y2EyMjY

**Comment by nicolas-grekas** [comment]
(the link should be active again - or try Twitter DM otherwise.)

**Comment by AshleyDawson** [comment]
@nicolas-grekas I've sent you a Twitter DM with a link to the file

**Comment by nicolas-grekas** [comment]
Thanks for the file, it's been definitely helpful. Fixed in #28760

**Comment by AshleyDawson** [comment]
No problem, thanks for resolving
